### DISCUSSION TITLE
How can I configure how the app starts?

### DISCUSSION THREAD

**Discussion by depesz** [jeremyevans/roda#277] [General]
First of all, sorry if I'm asking in the wrong place, but I'm not really developer, never wrote any webapp in ruby, but now am in position where I have to.

My whole "webapp" in roda is as simple as:

```
class App < Roda
  route do |r|
    r.post '' do
      response['Content-Type'] = 'application/json'
      PgQuery.parse(r['q']).tree.to_json
    end
  end
end
run App.freeze.app
```

And when I do "rackup", it works.

My question/problem is that i'd like to configure the port that it's listening on, make a pidfile and daemonize it.

I know I can do it with rackup --options, but was hoping for some configurability within ruby.

For sinatra apps I found that i can call "configure" method, and in it do stuff, but it doesn't seem to be the case with roda.

So, my question is: how can I configure the daemon behavior of roda apps? Is there an example/doc anywhere that would show me how to set ip/port/logfile/pidfile?

**Comment by jeremyevans** [comment]
First, this is the correct forum to ask questions.  Second, I'm a huge fan of your PostgreSQL blog posts, I've learned so much from them. So thank you for that.

Sinatra will automatically start a webserver in certain cases, which is why it offers features for configuring the web server it starts. Roda never starts a webserver, so it does not offer the API you are looking for. That's something you'll need to configure at the webserver level (i.e. using rackup/unicorn/puma/etc.).   If you are not familiar with the Ruby web ecosystem, that's probably more of a pain.

For now, I would advise using rackup or puma (rackup will use puma by default if puma is installed).  With rackup, you are looking for the `-o` (ip), `-p` (port), and `-P` (pid) options.  You'll want to redirect stdout/stderr for the logfile.  With puma, you are looking for the `-b` (bind, takes a URI), `-p` (port), and `--pidfile` (pid) options.  You can also redirect stdout/stderr on puma, or use the `--redirect-stdout`, `--redirect-stderr` , and `--redirect-append` options.

One other option to consider is the environment (`RACK_ENV`).  This defaults to `development` in both servers, but for a production application, you'll want to set it to `production`.  You can set the environment using `-E` in rackup and `-e` in puma. Note that in the production environment, I don't think either rackup or puma adds a default logging middleware.  Roda has a common_logger plugin that you can use for logging.

**Reply by depesz** [reply]
OK. So it looks that I should set all the params via options to rackup. And this means I'll use some shell scripting to avoid having to rememebr all that. That works.

I kinda suspect that 99% of my problems is unfamiliarity with the ecosystem, but the suspicion doesn't hjelp me solve my problem. Anyway. Will use shell script to control rackup, and it should be (i hope) OK.

Thanks :)
